contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,91 @@
+
+## Objective
+
+Provide a generic way to provide the local service route, say "topology aware routing of service". Local means "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc.
+
+## GOAL
+
+Support at least the following topology level
+
+* hostname
+
+* rack
+
+* zone
+
+* region
+
+## Non-goal
+
+ClusterIP. With NodePort and external LB, we may assume that some frontend already made a choice of which node to target, so staying node-local is valid. With ClusterIP, traffic is mostly assumed to be originating locally, so there will be clients on nodes that don't have backends for the services they are requesting.
+
+weight-based locality for now.
+
+## Use cases:
+
+* logging agents such as fluentd
+* authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy)
+* regional data costs in multi-AZ setup
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods. In this way, they can reduce network latency, improve security, save money and so on.
+
+Because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow more arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but ONLY for NodePort and external LB traffic. NodeName was added to EndpointAddress to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to reslove it.
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that we should not focus on the particular "node-local" user case, we should figure out a generic way.
+
+## Implementation details
+
+### API changes
+
+```go
+type Service struct {
+  spec:
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in node labels, e.g. kubernetes.io/hostname
+    // kubernetes.io/rack
+    // failure-domain.beta.kubernetes.io/zone
+    // failure-domain.beta.kubernetes.io/region
+    // ...
+  	TopologyKey string
+}
+```
+
+## Endpoint Controller Changes
+
+Endpoint Controller will filter out the backend pods based on the `TopologyKey`. Since the node labels is required, endpoint controller will need to watch nodes and maintain a key-value store for node cache - the key is the nodename and value is the node object. So, the new logic of endpoint controller might like:
+
+```go
+go wath service, pod, node
+if the TopologyKey in service is not empty; then
+	// filter out the pod that are NOT on the nodes whose label does NOT match the TopologyKey
```

Comment:
If I am reading it right, you suggest that if a node where a service's backend runs does not have appropriate label  then it will be completely excluded from the list of available  backends which means if by some reason local service's backend is not available, it would create this service outage. How about an "advisory" approach where local backends considers first, then next level and then default. Same approach as in the routing, more specific routes take precedence but if there is none, then default route saves you :)

Reply:
Yeah, it's an open question and deserves more discussion.